README.md
# JSPanel

A library that helps you write a static dropdown menu, a panel, that follows the digital accessibility recommendations.

## Get started

First of all, get the source code in the `src` folder of this repository. You need to copy the javascript file & the css file:

```html
<link rel="stylesheet" src="src/panel-style.css">
```

```html
<script src="src/JSPanel.js">
```

**Note:** the source code is available in TypeScript & JavaScript.

## Create your own panel

According to the digital accessibility recommendations, the panel needs to be opened by a button. Here is an example of what your code should be:

```html
<html>
<head>
  <style>
    #container {
      position: fixed;
      bottom: 30px;
      right: 30px;
    }
  </style>
</head>
<body>

  <div id="container">
    <button id="panel-controller">
  </div>

  <script src="src/JSPanel.js"></script>
  <script>
    const button = document.querySelector("#panel-controller");
    const panel = new JSPanel(button, {
      bottom: 0 + button.getBoundingClientRect().height,
      right: 0,
      items: [
        { title: "My Profile", onclick: () => console.log("et voilà") }
      ],
    });
  </script>
</body>
</html>
```

**Note:** the `aria-` attributes are added dynamically.

A panel has 5 different options:

|name|type|default value|description|
|----|----|-------------|-----------|
|`top`|number|null|The `top` CSS property.|
|`right`|number|null|The `right` CSS property.|
|`bottom`|number|null|The `bottom` CSS property.|
|`left`|number|null|The `left` CSS property.|
|`items`|Array of items|_required_|The items to be displayed in the panel.|

By default, the panel is placed in the upper left corner (`top:0,left:0`).

The items have also specific options:

|name|type|default value|description|
|----|----|-------------|-----------|
|`title`|string|(_required_)|The title of the item.|
|`icon`|string|null|The path to an image.|
|`fontawesome_icon`|string|null|The className of a Fontawesome icon.|
|`fontawesome_color`|string|null|The color of the Fontawesome icon.|
|`className`|string|null|Additional classes to be added to the item, separated by a white space.|
|`attributes`|`[string, string][]`|null|Additional attributes to be added to the item. Exemple: `[["data-thing", "thing"], ...[]]`|
|`onclick`|() => void|null|The callback function when the user clicks on the item.|
|`separator`|boolean|false|Displays a line. This item cannot have any other options.|

In order to use `fontawesome_icon` and `fontawesome_color`, make sure you've installed [Fontawesome](https://cdnjs.com/libraries/font-awesome) in your project.

## Customize the panel

You can change the style of the panel by modifying the CSS file. There are the main variables defined at the beginning of the file:

```css
:root {
    --panel-background-color: #fff;
    --panel-box-shadow: 0 0 4px rgba(204, 204, 204, 0.75);
    --panel-width: 160px;
    --panel-text-color: #404040;
    --panel-hover-item-background-color: #f4f6fa;
    --panel-hover-item-color: #385074;
    --panel-icon-width: 13px;
    --panel-separator-color: #dfe3eb;
}
```

For further modifications, you'll need to modify the CSS properties. Be careful not to modify the file too much at the risk of compromising the proper functioning of the panel.

## Digital Accessibility

Following the digital accessibility recommendations for this kind of panels, it is necessary to open or close the panel by clicking either the Enter or Space key. See <https://www.accede-web.com/en/guidelines/rich-interface-components/show-hide-panels/> for more information.

## License

MIT License

CHANGELOG.md
## [1.2.0] - April 28, 2021

New methods:

+ toggleItem
+ removeItem
+ removeItems
+ getAllIDS
+ getItem
+ addItem
+ replaceItemWith
+ deletePanel
+ isOpen

New option:

+ id

## [1.1.3] - April 28, 2021

Bug fixes

## [1.1.2] - April 25, 2021

Bug fixes

## [1.1.1] - April 23, 2021

Bug fixes

## [1.1.0] - April 23, 2021

Better digital accessibility.

## [1.0.2] - April 19, 2021

New options for the items:

* fontawesome_color
* className
* attributes

## [1.0.1] - April 15, 2021

Minor changes

## [1.0.0] - April 13, 2021

Initial release

src/JSPanel.js
"use strict";
/**
 * @class Creates a panel that follows the digital accessibility recommendations.
 */
class JSPanel {
    /**
     * @constructs JSPanel
     * @param {HTMLButtonElement} button The button which will display the panel.
     * @param {{top?:number,right?:number,bottom?:number,left?:number,items:Array<{title:string,icon?:string,fontawesome_icon?:string,fontawesome_color?:string,className?:string,attributes?:Array<Array<string>>,onclick?:Function,separator?:boolean}>}} options The options to customize the panel.
     */
    constructor(button, options) {
        /**
         * The panel.
         * @type {HTMLElement|null}
         * @default null
         * @private
         */
        this.panel = null;
        this.button = button;
        this.options = options;
        this.panel_uniqueid = "jspanel-" + this._rand(0, 1000000);
        this._buildPanel();
        this.button.setAttribute("aria-expanded", "false");
        this.button.setAttribute("aria-controls", this.panel_uniqueid);
    }
    /**
     * Builds the panel.
     * @private
     */
    _buildPanel() {
        const top = this.options.top === undefined ? null : this.options.top + "px";
        const right = this.options.right === undefined ? null : this.options.right + "px";
        const bottom = this.options.bottom === undefined ? null : this.options.bottom + "px";
        const left = this.options.left === undefined ? null : this.options.left + "px";
        this.panel = this._createEl("div", { id: this.panel_uniqueid, className: "jspanel panel-hidden" });
        if (top || right || bottom || left) {
            if (top)
                this.panel.style.top = top;
            if (left)
                this.panel.style.left = left;
            if (right)
                this.panel.style.right = right;
            if (bottom)
                this.panel.style.bottom = bottom;
        }
        else {
            this.panel.style.top = "0px";
            this.panel.style.left = "0px";
        }
        const parent = this.button.parentElement === null ? document.body : this.button.parentElement;
        const style_position = window.getComputedStyle(parent).getPropertyValue("position");
        if (!this._inArray(style_position, ["fixed", "absolute", "relative"]))
            parent.style.position = "relative";
        //
        // items
        //
        if (this.options.items) {
            const container = this._createEl("div", { className: "container-items" });
            for (let item of this.options.items) {
                if (item) {
                    const built_item = this._buildItem(item);
                    container.appendChild(built_item);
                }
            }
            this.panel.appendChild(container);
        }
        else {
            throw new Error("You need to define items to be displayed in the panel.");
        }
        //
        // events
        //
        document.addEventListener("click", (e) => {
            const target = e.target;
            if (target && this.panel) {
                if (!this.panel.contains(target) && this._isOpen()) {
                    this._closePanel();
                }
            }
        });
        this.button.onclick = (e) => { this._togglePanel(e); };
        this.button.onkeydown = (e) => { this._toggleOnKeyboardEvent(e); };
        this._insertAfterButton(this.panel);
        this.panel.onkeydown = (e) => {
            if (e.key === "Tab" || e.keyCode === 9) {
                if (this._isOpen())
                    this._focusInPanel(e);
            }
        };
        // I absolutly want the focus to stay inside (and only INSIDE) the panel :)
        // For that, I need to add a keydown event to the button too because
        // I want to include it during the keyboard navigation (with Tab)
        // So that it's easier for the user to close the panel with his/her keyboard.
        this.button.onkeydown = (e) => {
            if (e.key === "Tab" || e.keyCode === 9) {
                if (this._isOpen()) {
                    e.preventDefault();
                    const active_elements = this._getAllActiveItems();
                    if (active_elements && active_elements[0]) {
                        if (e.shiftKey === true) {
                            active_elements[active_elements.length - 2].focus(); // -2 because we don't want to take into account the button once again
                        }
                        else {
                            active_elements[0].focus();
                        }
                    }
                }
            }
        };
    }
    /**
     * Following the digital accessibility recommendations for this kind of panels,
     * it is necessary to open or close the panel by clicking either the Enter or Space key.
     * See {@link https://www.accede-web.com/en/guidelines/rich-interface-components/show-hide-panels/} for more information.
     * @param {KeyboardEvent} e The keyboard event.
     * @private
     */
    _toggleOnKeyboardEvent(e) {
        if (e.key === "Enter" || e.code === "Enter" || e.keyCode === 13 || e.key === " " || e.keyCode === 32 || e.code === "Space") {
            e.preventDefault();
            this._togglePanel(e);
        }
    }
    /**
     * Checks if the panel is currently opened or not.
     * @returns {boolean} True if the panel is opened.
     * @private
     */
    _isOpen() {
        if (this.panel) {
            return !this.panel.classList.contains("panel-hidden");
        }
        else {
            return false;
        }
    }
    /**
     * Open the panel if it's closed, close if it's opened.
     * @param {MouseEvent|KeyboardEvent} e The mouse event or the keyboard event.
     * @private
     */
    _togglePanel(e) {
        if (this.button && this.panel) {
            e.stopPropagation();
            if (this._isOpen()) {
                this._closePanel();
            }
            else {
                this.button.setAttribute("aria-expanded", "true");
                this.panel.classList.remove("panel-hidden");
                const all_items = this._getAllItems();
                if (all_items && all_items[0])
                    all_items[0].focus();
            }
        }
    }
    /**
     * Gets all the items from the panel if it's open.
     * @returns {NodeListOf<HTMLButtonElement>|null} All the items.
     */
    _getAllItems() {
        if (this._isOpen()) {
            return this.panel.querySelectorAll("button");
        }
        else {
            return null;
        }
    }
    /**
     * Gets all the active items from the panel if it's open.
     * @returns {Array<HTMLElement>|null} All the items that have an onclick property.
     */
    _getAllActiveItems() {
        if (this._isOpen()) {
            const active_elements = Array.from(this.panel.querySelectorAll("button"));
            active_elements.push(this.button);
            return active_elements;
        }
        else {
            return null;
        }
    }
    /**
     * Closes the panel.
     * @private
     */
    _closePanel() {
        if (this.button && this.panel) {
            this.button.setAttribute("aria-expanded", "false");
            this.panel.classList.add("panel-hidden");
        }
    }
    /**
     * Creates a customizable div element.
     * @param {string} tagName The name of the tag.
     * @param {{id?:string,className?:string,textContent?:string,attributes?:Array<string>,styles?:Array<string>}} options The options to customize the element.
     * @returns {HTMLElement} The created element.
     * @private
     */
    _createEl(tagName, options) {
        const el = document.createElement(tagName);
        if (!options)
            return el;
        if (options.id)
            el.setAttribute("id", options.id);
        if (options.textContent)
            el.textContent = options.textContent;
        if (options.className) {
            const classes = options.className.split(" ");
            for (let clas of classes) {
                el.classList.add(clas);
            }
        }
        if (options.styles) {
            for (let style of options.styles) {
                const property = style[0];
                const value = style[1];
                el.style[property] = value;
            }
        }
        if (options.attributes) {
            for (let attr of options.attributes) {
                const name = attr[0];
                const value = attr[1];
                el.setAttribute(name, value);
            }
        }
        return el;
    }
    /**
     * Builds an item.
     * @param {{title:string,icon?:string,fontawesome_icon?:string,fontawesome_color?:string,className?:string,attributes?:Array<Array<string>>,onclick?:Function,separator?:boolean}} item The item to build.
     * @returns {HTMLElement} The item as an HTML element.
     * @private
     */
    _buildItem(item) {
        if (item.separator) {
            const div = this._createEl("div", { className: 'jspanel-separator' });
            return div;
        }
        else {
            const button = this._createEl("button");
            button.setAttribute("aria-label", item.title);
            if ((item.icon && !item.fontawesome_icon) || (item.icon && item.fontawesome_icon)) {
                const icon = this._createEl("img", { attributes: [["src", item.icon]] });
                button.appendChild(icon);
            }
            else if (!item.icon && item.fontawesome_icon) {
                const icon = this._createEl("i", { className: item.fontawesome_icon });
                if (item.fontawesome_color)
                    icon.style.color = item.fontawesome_color;
                button.appendChild(icon);
            }
            if (item.className) {
                const classes = item.className.split(" ");
                for (let clas of classes) {
                    button.classList.add(clas);
                }
            }
            if (item.attributes) {
                for (let attr of item.attributes) {
                    const name = attr[0];
                    const value = attr[1];
                    button.setAttribute(name, value);
                }
            }
            const title = this._createEl("span", { textContent: item.title });
            button.appendChild(title);
            button.addEventListener('click', () => {
                if (item.onclick)
                    item.onclick();
                this._closePanel();
            });
            return button;
        }
    }
    /**
     * Blocks the focus inside the panel while it's open.
     * @param {KeyboardEvent} e The keyboard event.
     */
    _focusInPanel(e) {
        const all_items = this._getAllActiveItems();
        if (all_items) {
            e.preventDefault();
            let index = Array.from(all_items).findIndex(f => this.panel ? f === this.panel.querySelector(":focus") : false);
            if (e.shiftKey === true) {
                index--;
            }
            else {
                index++;
            }
            if (index >= all_items.length) {
                index = 0;
            }
            if (index < 0) {
                index = all_items.length - 1;
            }
            all_items[index].focus();
        }
    }
    /**
     * Checks the presence of an element in an array
     * @param {any} needle The element to search in the array.
     * @param {Array<any>} haystack The array in which to search for the element.
     * @param {boolean} strict Is the type of the needle necessary when searching? By default: false.
     * @returns {boolean} True if the needle was found.
     * @private
     */
    _inArray(needle, haystack, strict = false) {
        const length = haystack.length;
        for (let i = 0; i < length; i++) {
            if (strict) {
                if (haystack[i] === needle)
                    return true;
            }
            else {
                if (haystack[i] == needle)
                    return true;
            }
        }
        return false;
    }
    /**
     * Inserts the panel into the DOM (after the button).
     * @param {HTMLElement} panel The panel to insert after the button.
     * @private
     */
    _insertAfterButton(panel) {
        const parent = this.button.parentElement === null ? document.body : this.button.parentElement;
        parent.insertBefore(panel, this.button.nextSibling);
    }
    /**
     * Generates a random number [min;max[
     * @param {number} min The minimum number (included).
     * @param {number} max The maximum number (not included).
     * @returns {number} A random number between minimum and maximum.
     * @private
     */
    _rand(min, max) {
        min = Math.ceil(min);
        max = Math.floor(max);
        return Math.floor(Math.random() * (max - min)) + min;
    }
}


src/JSPanel.ts
interface PanelOptions {
    top?: number;
    right?: number;
    bottom?: number;
    left?: number;
    items: PanelItem[];
}

interface PanelItem {
    title: string;
    icon?: string;
    fontawesome_icon?: string;
    fontawesome_color?: string;
    className?: string;
    attributes?: [string, string][],
    onclick?: () => void;
    separator?: boolean;
}

interface PanelElement {
    id?: string;
    className?: string;
    textContent?: string;
    attributes?: [string, string][];
    styles?: [string, string][];
}

/**
 * @class Creates a panel that follows the digital accessibility recommendations.
 */
class JSPanel {
    /**
     * The panel.
     * @type {HTMLElement|null}
     * @default null
     * @private
     */
    private panel: HTMLElement | null = null;

    /**
     * The button which will display the panel.
     * @type {HTMLButtonElement}
     * @private
     */
    private button: HTMLButtonElement;

    /**
     * The options to customize the panel.
     * @type {{top?:number,right?:number,bottom?:number,left?:number,items:Array<{title:string,icon?:string,fontawesome_icon?:string,fontawesome_color?:string,className?:string,attributes?:Array<Array<string>>,onclick?:Function,separator?:boolean}>}}
     * @private
     */
    private options: PanelOptions;

    /**
     * The unique id of the panel.
     * @type {string}
     * @private
     */
    private panel_uniqueid: string;

    /**
     * @constructs JSPanel
     * @param {HTMLButtonElement} button The button which will display the panel.
     * @param {{top?:number,right?:number,bottom?:number,left?:number,items:Array<{title:string,icon?:string,fontawesome_icon?:string,fontawesome_color?:string,className?:string,attributes?:Array<Array<string>>,onclick?:Function,separator?:boolean}>}} options The options to customize the panel.
     */
    public constructor(button: HTMLButtonElement, options: PanelOptions) {
        this.button = button;
        this.options = options;
        this.panel_uniqueid = "jspanel-" + this._rand(0, 1000000);
        this._buildPanel();
        this.button.setAttribute("aria-expanded", "false");
        this.button.setAttribute("aria-controls", this.panel_uniqueid);
    }

    /**
     * Builds the panel.
     * @private
     */
    private _buildPanel(): void {
        const top = this.options.top === undefined ? null : this.options.top + "px";
        const right = this.options.right === undefined ? null : this.options.right + "px";
        const bottom = this.options.bottom === undefined ? null : this.options.bottom + "px";
        const left = this.options.left === undefined ? null : this.options.left + "px";

        this.panel = this._createEl("div", { id: this.panel_uniqueid, className: "jspanel panel-hidden" });
        if (top || right || bottom || left) {
            if (top) this.panel.style.top = top;
            if (left) this.panel.style.left = left;
            if (right) this.panel.style.right = right;
            if (bottom) this.panel.style.bottom = bottom;
        } else {
            this.panel.style.top = "0px";
            this.panel.style.left = "0px";
        }

        const parent = this.button.parentElement === null ? document.body : this.button.parentElement;
        const style_position = window.getComputedStyle(parent).getPropertyValue("position");
        if (!this._inArray(style_position, ["fixed", "absolute", "relative"])) parent.style.position = "relative";

        //
        // items
        //

        if (this.options.items) {
            const container = this._createEl("div", { className: "container-items" });
            for (let item of this.options.items) {
                if (item) {
                    const built_item = this._buildItem(item);
                    container.appendChild(built_item);
                }
            }
            this.panel.appendChild(container);
        } else {
            throw new Error("You need to define items to be displayed in the panel.");
        }

        //
        // events
        //

        document.addEventListener("click", (e) => {
            const target = e.target as HTMLElement;
            if (target && this.panel) {
                if (!this.panel.contains(target) && this._isOpen()) {
                    this._closePanel();
                }
            }
        });

        this.button.onclick = (e) => { this._togglePanel(e) };
        this.button.onkeydown = (e) => { this._toggleOnKeyboardEvent(e) };

        this._insertAfterButton(this.panel);

        this.panel.onkeydown = (e: KeyboardEvent) => {
            if (e.key === "Tab" || e.keyCode === 9) {
                if (this._isOpen()) this._focusInPanel(e);
            }
        };

        // I absolutly want the focus to stay inside (and only INSIDE) the panel :)
        // For that, I need to add a keydown event to the button too because
        // I want to include it during the keyboard navigation (with Tab)
        // So that it's easier for the user to close the panel with his/her keyboard.
        this.button.onkeydown = (e: KeyboardEvent) => {
            if (e.key === "Tab" || e.keyCode === 9) {
                if (this._isOpen()) {
                    e.preventDefault();

                    const active_elements = this._getAllActiveItems();
                    if (active_elements && active_elements[0]) {
                        if (e.shiftKey === true) {
                            active_elements[active_elements.length - 2].focus(); // -2 because we don't want to take into account the button once again
                        } else {
                            active_elements[0].focus();
                        }
                    }
                }
            }
        };
    }

    /**
     * Following the digital accessibility recommendations for this kind of panels,
     * it is necessary to open or close the panel by clicking either the Enter or Space key.
     * See {@link https://www.accede-web.com/en/guidelines/rich-interface-components/show-hide-panels/} for more information.
     * @param {KeyboardEvent} e The keyboard event.
     * @private
     */
    private _toggleOnKeyboardEvent(e: KeyboardEvent): void {
        if (e.key === "Enter" || e.code === "Enter" || e.keyCode === 13 || e.key === " " || e.keyCode === 32 || e.code === "Space") {
            e.preventDefault();
            this._togglePanel(e);
        }
    }

    /**
     * Checks if the panel is currently opened or not.
     * @returns {boolean} True if the panel is opened.
     * @private
     */
    private _isOpen(): boolean {
        if (this.panel) {
            return !this.panel.classList.contains("panel-hidden");
        } else {
            return false;
        }
    }

    /**
     * Open the panel if it's closed, close if it's opened.
     * @param {MouseEvent|KeyboardEvent} e The mouse event or the keyboard event.
     * @private
     */
    private _togglePanel(e: MouseEvent | KeyboardEvent): void {
        if (this.button && this.panel) {
            e.stopPropagation();

            if (this._isOpen()) {
                this._closePanel();
            } else {
                this.button.setAttribute("aria-expanded", "true");
                this.panel.classList.remove("panel-hidden");

                const all_items = this._getAllItems();
                if (all_items && all_items[0]) all_items[0].focus();
            }
        }
    }

    /**
     * Gets all the items from the panel if it's open.
     * @returns {NodeListOf<HTMLButtonElement>|null} All the items.
     */
    private _getAllItems(): NodeListOf<HTMLButtonElement> | null {
        if (this._isOpen()) {
            return (this.panel as HTMLElement).querySelectorAll("button");
        } else {
            return null;
        }
    }

    /**
     * Gets all the active items from the panel if it's open.
     * @returns {Array<HTMLElement>|null} All the items that have an onclick property.
     */
    private _getAllActiveItems(): HTMLElement[] | null {
        if (this._isOpen()) {
            const active_elements: HTMLElement[] = Array.from((this.panel as HTMLElement).querySelectorAll("button"));
            active_elements.push(this.button as HTMLElement);
            return active_elements;
        } else {
            return null;
        }
    }

    /**
     * Closes the panel.
     * @private
     */
    private _closePanel(): void {
        if (this.button && this.panel) {
            this.button.setAttribute("aria-expanded", "false");
            this.panel.classList.add("panel-hidden");
        }
    }

    /**
     * Creates a customizable div element.
     * @param {string} tagName The name of the tag.
     * @param {{id?:string,className?:string,textContent?:string,attributes?:Array<string>,styles?:Array<string>}} options The options to customize the element.
     * @returns {HTMLElement} The created element.
     * @private
     */
    private _createEl(tagName: string, options?: PanelElement): HTMLElement {
        const el = document.createElement(tagName);
        if (!options) return el;

        if (options.id) el.setAttribute("id", options.id);
        if (options.textContent) el.textContent = options.textContent;
        if (options.className) {
            const classes = options.className.split(" ");
            for (let clas of classes) {
                el.classList.add(clas);
            }
        }

        if (options.styles) {
            for (let style of options.styles) {
                const property = style[0];
                const value = style[1];
                el.style[property as any] = value;
            }
        }

        if (options.attributes) {
            for (let attr of options.attributes) {
                const name = attr[0];
                const value = attr[1];
                el.setAttribute(name, value);
            }
        }

        return el;
    }

    /**
     * Builds an item.
     * @param {{title:string,icon?:string,fontawesome_icon?:string,fontawesome_color?:string,className?:string,attributes?:Array<Array<string>>,onclick?:Function,separator?:boolean}} item The item to build.
     * @returns {HTMLElement} The item as an HTML element.
     * @private
     */
    private _buildItem(item: PanelItem): HTMLElement {
        if (item.separator) {
            const div = this._createEl("div", { className: 'jspanel-separator' });
            return div;
        } else {
            const button = this._createEl("button");
            button.setAttribute("aria-label", item.title);

            if ((item.icon && !item.fontawesome_icon) || (item.icon && item.fontawesome_icon)) {
                const icon = this._createEl("img", { attributes: [["src", item.icon]] });
                button.appendChild(icon);
            } else if (!item.icon && item.fontawesome_icon) {
                const icon = this._createEl("i", { className: item.fontawesome_icon });
                if (item.fontawesome_color) icon.style.color = item.fontawesome_color;
                button.appendChild(icon);
            }

            if (item.className) {
                const classes = item.className.split(" ");
                for (let clas of classes) {
                    button.classList.add(clas);
                }
            }

            if (item.attributes) {
                for (let attr of item.attributes) {
                    const name = attr[0];
                    const value = attr[1];
                    button.setAttribute(name, value);
                }
            }

            const title = this._createEl("span", { textContent: item.title });
            button.appendChild(title);
            
            button.addEventListener('click', () => {
                if (item.onclick) item.onclick();
                this._closePanel();
            });

            return button;
        }
    }

    /**
     * Blocks the focus inside the panel while it's open.
     * @param {KeyboardEvent} e The keyboard event.
     */
    private _focusInPanel(e: KeyboardEvent): void {
        const all_items = this._getAllActiveItems();
        if (all_items) {
            e.preventDefault();
            let index = Array.from(all_items).findIndex(f => this.panel ? f === this.panel.querySelector(":focus") : false);
            if (e.shiftKey === true) {
                index--;
            } else {
                index++;
            }
            if (index >= all_items.length) {
                index = 0;
            }
            if(index < 0) {
                index = all_items.length - 1;
            }
            all_items[index].focus();
        }
    }

    /**
     * Checks the presence of an element in an array
     * @param {any} needle The element to search in the array.
     * @param {Array<any>} haystack The array in which to search for the element.
     * @param {boolean} strict Is the type of the needle necessary when searching? By default: false.
     * @returns {boolean} True if the needle was found.
     * @private
     */
    private _inArray(needle: any, haystack: any[], strict: boolean = false): boolean {
        const length = haystack.length;
        for(let i = 0; i < length; i++) {
            if (strict) {
                if(haystack[i] === needle) return true;
            } else {
                if(haystack[i] == needle) return true;
            }
        }
        return false;
    }

    /**
     * Inserts the panel into the DOM (after the button).
     * @param {HTMLElement} panel The panel to insert after the button.
     * @private
     */
    private _insertAfterButton(panel: HTMLElement): void {
        const parent = this.button.parentElement === null ? document.body : this.button.parentElement;
        parent.insertBefore(panel, this.button.nextSibling);
    }

    /**
     * Generates a random number [min;max[
     * @param {number} min The minimum number (included).
     * @param {number} max The maximum number (not included).
     * @returns {number} A random number between minimum and maximum.
     * @private
     */
    private _rand(min: number, max: number): number {
        min = Math.ceil(min);
        max = Math.floor(max);
        return Math.floor(Math.random() * (max - min)) + min;
    }
}
